File: Computer-Architecture/cpu.c
#include <stdio.h>  // install libraries
#include <stdlib.h>
#include <string.h> // for memset string library
#include "cpu.h"

// #define DATA_LEN 6

// reading from the ram
unsigned char cpu_ram_read(struct cpu *cpu, unsigned char address)  // helper function; return the address of the cpu
{
  return cpu->ram[address]; // return whatever the address of the ram; the cpu at this address will return the bytes that's there
}

//write to the ram
void cpu_ram_write(struct cpu *cpu, unsigned char address, unsigned char value)  // helper function; pass in a value since we are writing to the ram; return the address of the cpu and value
{
  cpu->ram[address] = value; 
}

// Pushes a value on the CPU stack.
void cpu_push(struct cpu *cpu, unsigned char value)
{
  cpu->reg[SP]--; // decrement value 
  cpu_ram_write(cpu, cpu->reg[SP], value);  // call ram write function to push a value on to the stack
  // cpu_ram_write(cpu, cpu->reg[SP], cpu->reg[operandA]);  // call ram write function passing in cpu, address we need to push it to which is stored in the Stack Pointer, value of store
}

// Pops a value on the CPU stack.
unsigned char cpu_pop(struct cpu *cpu)
{
  unsigned char value = cpu_ram_read(cpu, cpu->reg[SP]); // pop copies the value from the address pointed to by SP to the given register.
  // cpu->reg[operandA] = cpu_ram_read(cpu, cpu->reg[SP]); 
  cpu->reg[SP]++; // increment the stack pointer 
  
  // unsigned char value = cpu->ram[cpu, cpu->reg[SP]];  

  return value;
}

/**
 * Load the binary bytes from a .ls8 source file into a RAM array
 */
void cpu_load(struct cpu *cpu, char *filename)  // cpu load taking in cpu and filename
{
  FILE *fp = fopen(filename, "r");  // open a file; grab the pointer to the file and open it
  // printf("%s\n", filename);

  char line[1024]; // give a buffer or enough memory for each line to scroll the whole line; 

  int address = cpu->pc;
  
/*
  char data[DATA_LEN] = {
    // From print8.ls8
    0b10000010, // LDI R0,8; 1 operand
    0b00000000,
    0b00001000,
    0b01000111, // PRN R0; 2 operand
    0b00000000,
    0b00000001  // HLT; 0 operand
  };

  int address = 0;  // starting address 

  for (int i = 0; i < DATA_LEN; i++) 
  {
    cpu->ram[address++] = data[i];
  }
*/

  // TODO: Replace this with something less hard-coded
  // int address = 0;  // starting address 

  // read the file
  if(fp == NULL)  // check first if the pointer is not Null
  {
    fprintf(stderr, "Cannot open file. %s\n", filename);  // if null, print error
    exit(2);  // exit with code 2
  }

  char *endchar;  // to initialize endchar

  // Read all the lines and store them in RAM
  // use string to line, strlon(), function to read one line at a time
  while(fgets(line, sizeof line, fp) != NULL)
  {
    unsigned char byte = strtol(line, &endchar, 2); // strtol; &endchar to ignore spaces and bad chars

    // ignore empty lines
    if(endchar == line) 
    {
      continue;
    }

    // storing the bytes in our ram
    cpu_ram_write(cpu, address++, byte); 
  }
}

/**
 * ALU
 */
// doing ALU operations; ALU for Arithmetic Logic Unit which is the fundamental building blocks of the cpu
void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB)
{
  switch (op) 
  {
    case ALU_MUL: // multiply
      // TODO
      // cpu->reg[regA] = cpu->reg[regA] * cpu->reg[regB];  // multiply values in register A and register B and store value in RegA
      cpu->reg[regA] *= cpu->reg[regB];
      break;

    // TODO: implement more ALU ops
    case ALU_ADD: // add      
      cpu->reg[regA] = cpu->reg[regA] + cpu->reg[regB]; 
      // cpu->reg[regA] += cpu->reg[regB]; 
      break;

    default:
      printf("Unknown ALU instruction at %02x: %02x\n", cpu->pc, op);
      exit(3);  // exit with code 3
  }
}

/**
 * Run the CPU
 */
void cpu_run(struct cpu *cpu)
{
  int running = 1; // True until we get a HLT instruction

  while (running) 
  {
    // TODO
    // 1. Get the value of the current instruction (in address PC).
    unsigned char IR = cpu_ram_read(cpu, cpu->pc); // instruction register; pass in cpu and the program count er to read off that; read from RAM at the PC; this is the instruction/ opcode

    // unsigned char operandA = cpu_ram_read(cpu, cpu->pc + 1); // openrandA will make 1 movement; reading at the next address of whatever the pc points at
    // unsigned char operandB = cpu_ram_read(cpu, cpu->pc + 2); // openrandB will make 2 movements; reading the next byte and store in operand B

    unsigned char operandA = cpu_ram_read(cpu, (cpu->pc + 1)& MAX_ADDR); // openrandA will make 1 movement; reading at the next address of whatever the pc points at
    unsigned char operandB = cpu_ram_read(cpu, (cpu->pc + 2)& MAX_ADDR); // openrandB will make 2 movements; reading the next byte and store in operand B


    // 2. switch() over it to decide on a course of action.
    // 3. Do whatever the instruction should do according to the spec.

    // printf("TRACE: %02x: %02x\n", cpu->pc, IR);
    // int instruction_set_pc = (IR >> 4) & 1;
    
    switch (IR) { // switch on the IR 

      case PRN: // call print function; print numeric value in the given register
        printf("%d\n", cpu->reg[operandA]); // print whatever is in operandA
        // printf("print 8 manually\n");        
        // cpu->pc += 2;  // move the program counter by 2
        break;
      
      case LDI:   // LDI/load instruction
        // printf("We're about to load.\n");
        cpu->reg[operandA] = operandB; // operandA tells us the register number we need to store the value in; operandB is the value we need to store 
        // cpu->pc += 3; // increment the pc; LDI is 3 bytes
        break;
      
      case HLT: // halt method
        // printf("This is the halt method.\n"); // does not move the program counter
        running = 0;  // then running is equal false or 0
        break;
      
      case MUL: 
        cpu->reg[operandA] *= cpu->reg[operandB]; // mutiplication 
        // alu(cpu, ALU_MUL, operandA, operandB);
        cpu->pc += 3; // increment the pc; LDI is 3 bytes      

        // cpu->reg[regA] = cpu->reg[regA] * cpu->reg[regB];
        
        break;

      case ADD: 
        cpu->reg[operandA] += cpu->reg[operandB]; // mutiplication 
        cpu->pc += 3; // increment the pc; LDI is 3 bytes      
        break;
      
      case PUSH:  // add push function 
        cpu_push(cpu, cpu->reg[operandA]);  
        // cpu->reg[SP]--;  // decrement the value
        // cpu_ram_write(cpu, cpu->reg[SP], cpu->reg[operandA]);  // call ram write function passing in cpu, address we need to push it to which is stored in the Stack Pointer, value of store
        break;
      
      case POP:
        cpu->reg[operandA] = cpu_pop(cpu);  // cpu_pop copies the value from the address pointed to by SP to the given register.
        // cpu->reg[operandA] = cpu_ram_read(cpu, cpu->reg[SP]); 
        // cpu->reg[SP]++; // increment the stack pointer 
        break;
      
      case PRA:
        printf("%c\n", cpu->reg[operandA]);
        break;

      case CALL:  // Call; 
        cpu_push(cpu, cpu->pc + 2); // push the address of the next instructions; call is 2 bytes long; next instruction after the call is pc + 2
        cpu->pc = cpu->reg[operandA];   // set the pc to a certain value
        break;

      case RET: // return 
        cpu->pc = cpu_pop(cpu); // pop the value off the stack and set the pc to it
        break;

// Sprting Challenge: 

      case CMP:   // Compares register A and register B.        
        if (cpu->reg[operandA] == cpu->reg[operandB]) // If they are equal, 
        {
          cpu->fl = 1;  // set the flag to 1.
        } else
        {
          cpu->fl = 0;  // Else set it to 0.
        }
        break;

      case JMP: // Jumps to the address stored in the given register.         
        cpu->pc = cpu->reg[operandA]; // Sets the PC to the address stored in the given register.
        break;

      case JEQ: // If equal (E) flag is true, jump to the address in the given register.
        if (cpu->fl)  
        {
          cpu->pc = cpu->reg[operandA];
        }
        else
        {
          cpu->pc += 2;
        }
        break;

      case JNE:           
        if (!cpu->fl) // If equal (E) flag is clear or false, jump to the address stored in the given register.
        {
          cpu->pc = cpu->reg[operandA];
        }
        else
        {
          cpu->pc += 2;
        }
        break;
      
      default:  // default case; if something wrong happes; will catch anything that it doesn't know what to do
        printf("Unknown instruction at %02x: %02x\n", cpu->pc, IR); // if we get instruction doesn't know, this will tell us where and what it is; tell it the pc counter which it's pointing at
        exit(2);  // exit with code 2
    }

    // 4. Move the PC to the next instruction.
    // cpu->pc += (IR >> 6) + 1;

    int instruction_set_pc = (IR >> 4) & 1; // initialize instruction_set_pc

    if (!instruction_set_pc)  // if not 
    {
      cpu->pc += ((IR >> 6) & 0x3) + 1; // Shifts 6 places to the right.
    }
  }
}

/**
 * Initialize a CPU struct
 */
void cpu_init(struct cpu *cpu)
{
  // TODO: Initialize the Program Counter (PC) and other special registers
  // cpu->pc = 0;  // initialize/ load instructions at address 0
  cpu->pc = START_OF_PROGRAM_ADDR;  // initialize/ set to start of program address
  cpu->fl = CPU_FLAG;     // flag

  // cpu->reg[SP] = 0xf4;  // setup register; SP points at address F4 if the stack is empty

  cpu->reg[SP] = START_OF_STACK_ADDR; // Initialize the Stack Pointer (SP).
  cpu->reg[IM] = INTERRUPT_MASK;      // Initialize the Other special registers.
  cpu->reg[IS] = INTERRUPTS;

  // TODO: Zero registers and RAM
  memset(cpu->reg, 0, sizeof cpu->reg);  // grab the registers; we fill in the array with zeros
  memset(cpu->ram, 0, sizeof cpu->ram);  // grab the ram; we fill in the array with zeros
}

File: Computer-Architecture/ls8/cpu.py
"""CPU functionality."""

import sys

SP = 7

HLT = 0b00000001
LDI = 0b10000010
PRN = 0b01000111
MUL = 0b10100010
PUSH = 0b01000101
POP = 0b01000110
NOP = 0b00000000

#### SPRINT CHALLENGE ####
CMP = 0b10100111
JEQ = 0b01010101
JMP = 0b01010100
JNE = 0b01010110

class CPU:
    """Main CPU class."""  

    def __init__(self):
        """Construct a new CPU."""
        # init 8 registers
        self.reg = [0] * 8
        # we increment pc--it's the index of the current instructions
        self.pc = 0
        # Init memory with 256 bits
        self.ram = [0b0] * 256
        # Per our spec, reg 7 = 0xF4
        self.reg[7] = 0xF4

        self.fl = None

        # TODO: Branch table 
        self.branch_table = {
            HLT : self.hlt,
            LDI : self.ldi,
            PRN : self.prn,
            MUL : self.mul,
            PUSH : self.push,
            POP : self.pop,
            NOP : self.nop,
            CMP : self.cmp,
            
            #### SPRINT CHALLENGE ####
            JEQ : self.jeq,
            JMP : self.jmp,
            JNE : self.jne
        }

    ##### SPRINT CHALLENGE #####

    def cmp(self, operand_a, operand_b):
        # change flag depending on opA and opB
        if self.reg[operand_a] == self.reg[operand_b]:
            self.fl = "E"
        elif self.reg[operand_a] < self.reg[operand_b]:
            self.fl = "LT"
        elif self.reg[operand_a] > self.reg[operand_b]:
            self.fl = "GT"
        else:
            self.fl = 0

        self.pc += 3

    def jeq(self, operand_a, operand_b):
        # if flag is E (equal) jump to address stored in given register
        if self.fl == "E":
            self.pc = self.reg[operand_a]
        else:
            self.pc += 2
            
    def jmp(self, operand_a, operand_b):
        self.pc = self.reg[operand_a]

    def jne(self, operand_a, operand_b):
        if self.fl != "E":
            self.pc = self.reg[operand_a]
        else:
            self.pc += 2

    ############################
   
     # HLT, or Halt: exits the emulator
    def hlt(self, operand_a, operand_b):
        sys.exit(0)

    # LDI, or Load Immediate; set specified register to specific value
    def ldi(self, operand_a, operand_b):
        self.reg[operand_a] = operand_b
        # Increment program counter by 3 steps in RAM
        self.pc += 3

    # PRN, or print register: prints the current register
    def prn(self, operand_a, operand_b):
        print(self.reg[operand_a])
        self.pc += 2

    # Mul, or multiply: multiplies the next two values.
    def mul(self, operand_a, operand_b):
        self.reg[operand_b] *= self.reg[operand_b]
        self.pc += 3

    #PUSH: Push the value to the stack, decrement the pointer
    def push(self, operand_a, operand_b):
        # decrement SP
        self.reg[SP] -= 1
        # Get the value we want to store from the register and store it in ram
        self.ram_write(self.reg[operand_a], self.reg[SP])
        self.pc += 2

    # POP: pop the top value off the stack
    def pop(self, operand_a, operand_b):
        # get the value of SP and overwrite next register
        value = self.ram_read(self.reg[SP])
        self.reg[operand_a] = value
        # increment SP
        self.reg[SP] += 1
        self.pc += 2

    # NOP, or NOPE: passes when the value given is 0s
    def nop(self, operand_a, operand_b):
        self.pc += 1

    # RET, or RETURN: returns back to where we called from
    def ret(self, operand_a, operand_b):
        self.pc = self.ram[self.reg[SP]]
        self.reg[SP] += 1 

    # CALL: jump to any adress we specify
    def call(self, operand_a, operand_b):
        self.reg[SP] -= 1
        self.ram_write(self.pc + 2, self.reg[SP])
        self.pc = self.reg[operand_a]  
 

    def load(self, filename=None):
        """Load a program into memory."""

        address = 0

        # For now, we've just hardcoded a program:
        if filename:
            with open('./examples/' + filename) as f:
                address = 0
                for line in f:
                    line = line.split("#")[0].strip()
                    if line == '':
                        continue
                    else:
                        instruction = int(line, 2)  # by default this is base 10 so we need to change this to parse base 2 number
                        self.ram[address] = instruction  
                        address += 1         

        else:
            program = [
                # From print8.ls8
                0b10000010, # LDI R0,8
                0b00000000,
                0b00001000,
                0b01000111, # PRN R0
                0b00000000,
                0b00000001, # HLT
            ]

            for address, instruction in enumerate(program):
                self.ram[address] = instruction

    def alu(self, op, reg_a, reg_b):
        """ALU operations."""

        if op == "ADD":
            self.reg[reg_a] += self.reg[reg_b]
        elif op == "SUB": 
            self.reg[reg_a] -= self.reg[reg_b]
        elif op == "MUL":
            self.reg[reg_a] *= self.reg[reg_b]
        else:
            raise Exception("Unsupported ALU operation")

    def ram_read(self, address):
        """prints what's stored in that specified address in RAM"""
        return self.ram[address]

    def ram_write(self, value, address):
        """Overwrites the address in ram with the value"""
        self.ram[address] = value

    def trace(self):
        """
        Handy function to print out the CPU state. You might want to call this
        from run() if you need help debugging.
        """

        print(f"TRACE: %02X | %02X %02X %02X |" % (
            self.pc,
            #self.fl,
            #self.ie,
            self.ram_read(self.pc),
            self.ram_read(self.pc + 1),
            self.ram_read(self.pc + 2)
        ), end='')

        for i in range(8):
            print(" %02X" % self.reg[i], end='')

        print()

    def run(self):
        """Run the CPU."""

        while True:
            # self.trace()
            op_code = self.ram[self.pc]
            operand_a, operand_b  = self.ram[self.pc + 1], self.ram[self.pc + 2]
            if op_code in self.branch_table:
                self.branch_table[op_code](operand_a, operand_b)
            else: 
                self.pc += 1

if __name__ == '__main__':
    LS8 = CPU()
    LS8.load()
    for i in range(9):
        print(LS8.ram_read(i))
    LS8.ram_write(0, 15)
    print('==============')
    print(LS8.ram_read(0))
    print('==============')